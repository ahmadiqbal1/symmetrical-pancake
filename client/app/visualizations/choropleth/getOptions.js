import { isNil, merge, first, keys, get } from "lodash";
import ColorPalette from "./ColorPalette";
import availableMaps from "./maps";

const defaultMap = first(keys(availableMaps));

const DEFAULT_OPTIONS = {
  mapType: defaultMap,
  customMapUrl: null,
  keyColumn: null,
  targetField: null,
  valueColumn: null,
  clusteringMode: "e",
  steps: 5,
  valueFormat: "0,0.00",
  noValuePlaceholder: "N/A",
  colors: {
    min: ColorPalette["Light Blue"],
    max: ColorPalette["Dark Blue"],
    background: ColorPalette.White,
    borders: ColorPalette.White,
    noValue: ColorPalette["Light Gray"],
  },
  legend: {
    visible: true,
    position: "bottom-left",
    alignText: "right",
  },
  tooltip: {
    enabled: true,
    template: "<b>{{ @@name }}</b>: {{ @@value }}",
  },
  popup: {
    enabled: true,
    template: "Country: <b>{{ @@name_long }} ({{ @@iso_a2 }})</b>\n<br>\nValue: <b>{{ @@value }}</b>",
  },
};

export default function getOptions(options) {
  const result = merge({}, DEFAULT_OPTIONS, options);
  // Both renderer and editor always provide new `bounds` array, so no need to clone it here.
  // Keeping original object also reduces amount of updates in components
  result.bounds = get(options, "bounds");

  if (isNil(availableMaps[result.mapType]) && result.mapType !== "custom") {
    result.mapType = defaultMap;
  }

  // backward compatibility
  if (!isNil(result.countryCodeColumn)) {
    result.keyColumn = result.countryCodeColumn;
    delete result.countryCodeColumn;
  }
  if (!isNil(result.countryCodeType)) {
    result.targetField = result.countryCodeType;
    delete result.countryCodeType;
  }

  return result;
}
