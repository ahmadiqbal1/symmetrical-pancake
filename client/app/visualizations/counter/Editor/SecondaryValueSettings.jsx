import React from "react";
import { Section, Checkbox } from "@/components/visualizations/editor";
import { EditorPropTypes } from "@/visualizations/prop-types";
import CounterValueOptions from "./CounterValueOptions";
import counterTypes from "../counterTypes";

export default function SecondaryValueSettings({ options, data, onOptionsChange }) {
  const onChange = secondaryValue => onOptionsChange({ secondaryValue });

  const disabled = options.secondaryValue.type === "unused";

  return (
    <React.Fragment>
      <CounterValueOptions
        disabled={disabled}
        counterTypes={counterTypes}
        options={options.secondaryValue}
        data={data}
        onChange={onChange}
      />
      <Section>
        <Checkbox
          data-test="Counter.ShowSecondaryValue"
          disabled={disabled}
          checked={options.secondaryValue.show}
          onChange={e => onChange({ show: e.target.checked })}>
          Show Secondary Value
        </Checkbox>
      </Section>
    </React.Fragment>
  );
}

SecondaryValueSettings.propTypes = EditorPropTypes;
