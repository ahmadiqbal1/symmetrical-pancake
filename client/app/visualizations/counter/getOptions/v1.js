import { extend } from "lodash";

const schemaVersion = 1;

const defaultOptions = {
  counterLabel: "",
  counterColName: "counter",
  rowNumber: 1,
  targetColName: null,
  targetRowNumber: 1,
  stringDecimal: 0,
  stringDecChar: ".",
  stringThouSep: ",",
  stringPrefix: null,
  stringSuffix: null,
  formatTargetValue: false,
  countRow: false,
};

export default function getOptions(options) {
  return extend({}, defaultOptions, options, { schemaVersion });
}
