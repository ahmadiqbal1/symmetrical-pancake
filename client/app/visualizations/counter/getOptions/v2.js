import { isEmpty, isFinite, merge, get, pick, padEnd, toString, includes, map } from "lodash";
import getOptionsV1 from "./v1";

const schemaVersion = 2;

const defaultOptions = {
  schemaVersion,

  counterLabel: "",

  stringDecChar: ".",
  stringThouSep: ",",
  numberFormat: "0,0",

  primaryValue: {
    show: true,
    type: "rowValue",
    column: "counter",
    rowNumber: 1,
    displayFormat: "{{ @@value_formatted }}",
    showTooltip: true,
    tooltipFormat: "{{ @@value }}",
  },
  secondaryValue: {
    show: true,
    type: "unused",
    column: null,
    rowNumber: 1,
    displayFormat: "({{ @@value_formatted }})",
    showTooltip: true,
    tooltipFormat: "{{ @@value }}",
  },
};

function migrateFromV1(options) {
  options = getOptionsV1(options);
  const result = pick(options, ["counterLabel", "stringDecChar", "stringThouSep"]);

  result.numberFormat = "0,0.000";
  if (isFinite(options.stringDecimal) && options.stringDecimal >= 0) {
    result.numberFormat = "0,0";
    if (options.stringDecimal > 0) {
      const decimals = padEnd("", options.stringDecimal, "0");
      result.numberFormat = `${result.numberFormat}.${decimals}`;
    }
  }

  const prefix = toString(options.stringPrefix);
  const suffix = toString(options.stringSuffix);

  result.primaryValue = {
    show: true,
    type: options.countRow ? "countRows" : "rowValue",
    column: options.counterColName,
    rowNumber: options.rowNumber,
    displayFormat: `${prefix}{{ @@value_formatted }}${suffix}`,
  };

  result.secondaryValue = {
    show: true,
    type: options.targetColName ? "rowValue" : "unused",
    column: options.targetColName,
    rowNumber: options.targetRowNumber,
    displayFormat: options.formatTargetValue ? `(${prefix}{{ @@value_formatted }}${suffix})` : "({{ @@value }})",
  };

  return result;
}

export default function getOptions(options, { columns }) {
  const currentSchemaVersion = get(options, "schemaVersion", isEmpty(options) ? schemaVersion : 0);
  if (currentSchemaVersion < schemaVersion) {
    options = migrateFromV1(options);
  }

  const optionsWithDefaults = merge({}, defaultOptions, options, { schemaVersion });
  const columnNameUpdates = {};

  const columnNames = map(columns, col => col.name);
  if (!includes(columnNames, get(optionsWithDefaults, "primaryValue.column"))) {
    columnNameUpdates.primaryValue = { column: null };
  }
  if (!includes(columnNames, get(optionsWithDefaults, "secondaryValue.column"))) {
    columnNameUpdates.secondaryValue = { column: null };
  }

  return merge({}, optionsWithDefaults, columnNameUpdates);
}
