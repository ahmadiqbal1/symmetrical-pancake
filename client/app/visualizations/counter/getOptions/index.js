import getOptions from "./v2";

export default getOptions;
