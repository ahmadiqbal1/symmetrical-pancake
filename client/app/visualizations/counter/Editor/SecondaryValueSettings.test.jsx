import { after } from "lodash";
import React from "react";
import enzyme from "enzyme";

import getOptions from "../getOptions";
import SecondaryValueSettings from "./SecondaryValueSettings";

function findByTestID(wrapper, testId) {
  return wrapper.find(`[data-test="${testId}"]`);
}

function mount(options, done) {
  const data = {
    columns: [
      { name: "a", type: "number" },
      { name: "b", type: "number" },
    ],
    rows: [{ a: 123, b: 987 }],
  };

  options = getOptions(options, data);
  return enzyme.mount(
    <SecondaryValueSettings
      visualizationName="Test"
      data={data}
      options={options}
      onOptionsChange={changedOptions => {
        expect(changedOptions).toMatchSnapshot();
        done();
      }}
    />
  );
}

describe("Visualizations -> Counter -> Editor -> Primary Value Settings", () => {
  test("Changes Counter Type", done => {
    const el = mount({ schemaVersion: 2, primaryValue: { type: "unused" } }, done);

    findByTestID(el, "Counter.CounterType")
      .last()
      .simulate("click");
    findByTestID(el, "Counter.CounterType.rowValue")
      .last()
      .simulate("click");
  });

  test("Changes Counter Type Options", done => {
    // we will perform 2 actions, so call `done` after all of them completed
    const el = mount({ schemaVersion: 2, secondaryValue: { type: "rowValue" } }, after(2, done));

    findByTestID(el, "Counter.ColumnName")
      .last()
      .simulate("click");
    findByTestID(el, "Counter.ColumnName.a")
      .last()
      .simulate("click");

    findByTestID(el, "Counter.RowNumber")
      .last()
      .find("input")
      .simulate("change", { target: { value: "3" } });
  });

  test("Changes Format Options", done => {
    // we will perform 3 actions, so call `done` after all of them completed
    const el = mount({ schemaVersion: 2, secondaryValue: { type: "rowValue" } }, after(3, done));

    findByTestID(el, "Counter.DisplayFormat")
      .last()
      .find("input")
      .simulate("change", { target: { value: "{{ @@value_formatted }} ({{ @@value }})" } });

    findByTestID(el, "Counter.ShowTooltip")
      .last()
      .find("input")
      .simulate("change", { target: { checked: false } });

    findByTestID(el, "Counter.TooltipFormat")
      .last()
      .find("input")
      .simulate("change", { target: { value: "{{ @@value_formatted }} / {{ @@value }}" } });
  });

  test("Hides Secondary Value", done => {
    const el = mount({ schemaVersion: 2, secondaryValue: { show: true, type: "rowValue" } }, done);

    findByTestID(el, "Counter.ShowSecondaryValue")
      .last()
      .find("input")
      .simulate("change", { target: { checked: false } });
  });
});
