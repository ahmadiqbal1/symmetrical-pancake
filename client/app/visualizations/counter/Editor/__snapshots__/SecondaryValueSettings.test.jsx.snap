// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visualizations -> Counter -> Editor -> Primary Value Settings Changes Counter Type 1`] = `
Object {
  "secondaryValue": Object {
    "type": "rowValue",
  },
}
`;

exports[`Visualizations -> Counter -> Editor -> Primary Value Settings Changes Counter Type Options 1`] = `
Object {
  "secondaryValue": Object {
    "column": "a",
  },
}
`;

exports[`Visualizations -> Counter -> Editor -> Primary Value Settings Changes Counter Type Options 2`] = `
Object {
  "secondaryValue": Object {
    "rowNumber": 3,
  },
}
`;

exports[`Visualizations -> Counter -> Editor -> Primary Value Settings Changes Format Options 1`] = `
Object {
  "secondaryValue": Object {
    "displayFormat": "{{ @@value_formatted }} ({{ @@value }})",
  },
}
`;

exports[`Visualizations -> Counter -> Editor -> Primary Value Settings Changes Format Options 3`] = `
Object {
  "secondaryValue": Object {
    "tooltipFormat": "{{ @@value_formatted }} / {{ @@value }}",
  },
}
`;

exports[`Visualizations -> Counter -> Editor -> Primary Value Settings Hides Secondary Value 1`] = `
Object {
  "secondaryValue": Object {
    "show": false,
  },
}
`;

exports[`Visualizations -> Counter -> Editor -> Primary Value Settings Changes Format Options 2`] = `
Object {
  "secondaryValue": Object {
    "showTooltip": false,
  },
}
`;
